### ISSUE TITLE
storage: chunk gcqueue's requests

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#15989]
The GC queue may generate large proposals, especially now that proposer-evaluated KV is in place. Though it hasn't been observed in practice, it should be easy to trigger that by deleting everything in a full range at around the same time (and we should start doing that on our test clusters).

The GC Queue should simply dispatch its `GCRequest` batches in chunks. Since it scans the complete key range before, it can even be more intelligent about these chunks than just capping them at X keys (which may still incur a large proposal; think of the case in which one key is continuously overwritten - that would generate a huge proposal just out of the evaluation of the GCRequest affecting only that key).

**Comment by tschottdorf** [comment]
We've seen in #16068 that keys with 1000+ old versions pose a problem for `MVCCGarbageCollect` due to the large number of engine operations. For example, with 1000 keys and 1000 versions each, we clock in at ~3s in my laptop.

These numbers are not unlikely to pop up in `UPSERT`-heavy workloads in practice., though implementing #7880 along with sending a separate RPC for the threshold bumps should make these requests not block other traffic as much as they do at the time of writing.

**Comment by tschottdorf** [comment]
The current status is that we're seeing ~20k key requests in production and they're not so bad, taking a few milliseconds. Still, there might be some benefit to batching more here.

**Comment by tschottdorf** [comment]
I'll leave this as an external contributor / starter issue. The WriteBatches created by GC contain only `engine.Clear`, so it should be really difficult to get them large (right @petermattis?)

**Comment by petermattis** [comment]
The WriteBatches will contain every GC'd key. Depending on your table schema, keys may contain a significant fraction of the data.

@a6802739 this might be something good for you to tackle.

**Comment by a6802739** [comment]
@petermattis, okay, I could try to tackle this.

**Comment by tschottdorf** [comment]
Thanks! The code lives in `gc_queue.go`. Seeing that the size of the created proposals should depend linearly on the size of the keys, you could track the key bytes and batch every time this exceeds a threshold (so that we roughly send X bytes of keys for GC in each batch).

**Comment by a6802739** [comment]
thank you   very   much!

**Comment by a6802739** [comment]
@tschottdorf, So we need to keep track of the size of `gcKeys`, and we should chunk the `GCRequest ` when the size of the `gcKeys` exceeds the threshold?

So how should we set the  `X` bytes of keys, through a global variable or the environment variable? And for now, what value we should set `X` to ?

**Comment by tschottdorf** [comment]
@a6802739 exactly, where "size" would be something like `for _, key := range gcKeys { size += len(key.Key) }`. I think for now you can hardcode a size in this file; I'd suggest something like 256kb to get started.

**Comment by a6802739** [comment]
@tschottdorf, got it. Thank you very much.
